### ISSUE TITLE
ISC003 gives false positive when two different string types in one line are mandatory

### ISSUE THREAD

**Issue by mmarras** [astral-sh/ruff#6595]
When defining `hovertemplates` they use a magic string (e.g. `%{x}` to denote the x-data)  that cannot be inside an fstring. If one wants to combine the former with an fstring on the same line it flags the thing and it is impossible to comply to ruff ISC without breaking the code. So even if one decides to keep the rule for same line, it should make exception for mixed string types.

Correct : 
```python
import plotly.graph_objs as go
string = "Data1"
go.Figure(data=go.Scatter(
    x=[1.123123,2.12321,3.12321],
    y=[1.12321,2.12321,3.12321],
    hovertemplate="%{x:.2f}<br>%{y:.3f}" + f"<extra>{string}</extra>")
          )
```
The above codeblock but with ` hovertemplate=f"%{x:.2f}<br>%{y:.3f}<extra>{string}</extra>")` fails.
> NameError: name 'x' is not defined

Of course the correct version (see codeblock) triggers 
> Explicitly concatenated string should be implicitly concatenatedRuff[ISC003](https://beta.ruff.rs/docs/rules/explicit-string-concatenation)

or when implicitly concatenated:
> Implicitly concatenated string literals on one lineRuff[ISC001](https://beta.ruff.rs/docs/rules/single-line-implicit-string-concatenation)

_Originally posted by @mmarras in https://github.com/astral-sh/ruff/issues/5332#issuecomment-1657277651_

**Comment by charliermarsh** [comment]
Having trouble reproducing.

Given:
```python
import plotly.graph_objs as go
string = "Data1"
go.Figure(data=go.Scatter(
    x=[1.123123,2.12321,3.12321],
    y=[1.12321,2.12321,3.12321],
    hovertemplate="%{x:.2f}<br>%{y:.3f}" + f"<extra>{string}</extra>")
          )
```

Running `ruff check foo.py --select ISC --isolated -n` doesn't yield any violations.

Was it broken over multiple lines? Like:

```python
import plotly.graph_objs as go
string = "Data1"
go.Figure(data=go.Scatter(
    x=[1.123123,2.12321,3.12321],
    y=[1.12321,2.12321,3.12321],
    hovertemplate=(
        "%{x:.2f}<br>%{y:.3f}" +
        f"<extra>{string}</extra>")
    )
)
```

**Comment by mmarras** [comment]
Indeed, this only shows in my VScode (1.81.0) extension (0.0.277, I assume), not if I run ruff cli with ruff (0.0.284). My bad.

 
![image](https://github.com/astral-sh/ruff/assets/53915566/59cc6cf5-1334-48e1-8d50-8843c491945a)

**Comment by zanieb** [comment]
Hm I think the VSCode extension should use your externally installed Ruff if available (so an update is not needed to get the latest version). Anyway, it sounds like we've resolved this on the latest version. Let us know if there's more we can do!

**Comment by charliermarsh** [comment]
I think it's not quite resolved because the request is that we _don't_ flag explicit concatenations with mixed-kind strings, like:

```python
import plotly.graph_objs as go
string = "Data1"
go.Figure(data=go.Scatter(
    x=[1.123123,2.12321,3.12321],
    y=[1.12321,2.12321,3.12321],
    hovertemplate=(
        "%{x:.2f}<br>%{y:.3f}" +
        f"<extra>{string}</extra>")
    )
)
```

**Comment by charliermarsh** [comment]
(The confusion above is about whether it flags for multiline strings or single-line strings, related to the repro.)

**Comment by mmarras** [comment]
> Hm I think the VSCode extension should use your externally installed Ruff if available (so an update is not needed to get the latest version). Anyway, it sounds like we've resolved this on the latest version. Let us know if there's more we can do!

So, VScode was in a different env than my terminal conda env, therefore the missmatch in ruff results editor vs cli. 

As far as I am concerned **the issue is solved** (I checked, as of 0.0.281). Conclusion stands: my bad. 

For me it was always about the single line example. I was happy as soon as an explicit concat of  mixed-kind one-line string wasn't flagged anymore. But I'm using plotly quite a bit so, if it comes up in another constellation, I can for sure report back.

**Comment by charliermarsh** [comment]
Okay, sounds good! I'm glad it's resolved for this use-case, thank you for reporting back :)
